joomla/joomla-cms
# Client-side JText.plural() implementation & [Admin Dashboard] Make Extension Updater display single update-availability correctly

[EDIT]
### Summary of Changes
- This PR now includes a first stab at introducing a server-assisted client-side JText.plural() implementation.
The initial reason of this PR, is also resolved in the same PR.

- Makes Extension Updater display single and multiple update-availability-text correctly in Dashboard (Control Panel):

When a single update is found:
 - the sidebar will now show "Update is Available [1]"
 - the message container will show "[1] Extension Update is available: [View Update]"

When multiple updates are found:
 - the sidebar will now show "Updates are Available [x]"
 - the message container will show "[x] Extension Updates are available: [View Updates]"


Note: This PR now has a proposed implementation of a javascript JText.plural method, and I'd like to read your thoughts on this proposal.

Implementation details to allow for client side plural implementation:
- The proposed changes should maintain B/C, as in existing functionality should not break. The extension-updater has been updated in this same PR, to work with either patched, or unpatched language files (*localise.php).
- I introduced a new method in language/localise.php files, which provides the necessary data and JS-code, for the client pluralization to work.
- In the back-end I introduced a JText::pluralJS() method, which gathers all the wanted pluralized language strings and pushes them to the frontend via the Javascript option storage.
- In the front-end I introduced the Joomla.Jtext.plural() method, which accepts the wanted language string and the quantifier number. It will then return the pluralized string, based on all the above.

Thoughts & comments, welcome.



### Testing Instructions

- Code Review

- Apply patch

- Test the following:

The tests below should be tested each against none, a single and multiple updates for extensions.
For the tests I have included some zipped files here: [testfiles.zip](https://github.com/joomla/joomla-cms/files/681200/testfiles.zip)
This zip file includes 4 files
- two patched `*localise.php` files, one for French and one for Russian.
- two patched extensionupdater language files again for French and Russian

Note: To avoid any false errors, it is best to disable the browser cache.

1) Test pluralization with the new en-GB pluralizers. (files are ready)
Result: The new client side pluralization should show the correct pluralization.

2) Test in other languages where the localise.php files have not been patched.
Result: The new client side pluralization should still show results the old way (not correctly pluralized).

3) Test pluralization and fallback to en-GB for other languages. For this test please use one or both of the patched `*localise.php` files. Please copy to the appropriate language directories in admin and site.
Result: The new client side pluralization should still show correct pluralization with a fallback to en-GB strings.
Note: If testing with the Russian language, note that en-GB does not have the `_2` suffix, so there is no fallback and the language string will be presented.

4) Test pluralization with patched extensionupdater language files. I have patched them both to display [FR] with the equivalent English text and [RU*] where * is replaced with the suffixes generated. Please put those in the appropriate directory in administrator/languages.
Result: It should correctly display the pluralized strings for that language.


### Documentation Changes Required

[EDIT] Yes, as for the language files `*localise.php` and the description of the new functionality.
